src/sqlancer/StatementReducer.java

```diff
@@ -28,32 +33,59 @@ public void reduce(G state, Reproducer<G> reproducer, G newGlobalState) throws E
         printQueries(knownToReproduceBugStatements);
         System.out.println();
 
-        do {
+        if (knownToReproduceBugStatements.size() <= 1) {
+            return;
+        }
+
+        partitionNum = 2;
+        while (knownToReproduceBugStatements.size() >= 2) {
             observedChange = false;
+
             knownToReproduceBugStatements = tryReduction(state, reproducer, newGlobalState,
-                    knownToReproduceBugStatements, (candidateStatements, i) -> {
-                        candidateStatements.remove((int) i);
+                    knownToReproduceBugStatements, (candidateStatements, s, l) -> {
+                        List<Query<C>> removed = new ArrayList<>();
+                        for (int i = 0; i < candidateStatements.size(); i++) {
+                            if (i >= s + l || i < s) {
+                                removed.add(candidateStatements.get(i));
+                            }
+                        }
+                        candidateStatements.clear();
+                        candidateStatements.addAll(removed);
                         return true;
                     });
-        } while (observedChange);
+            if (!observedChange) {
+                if (partitionNum == knownToReproduceBugStatements.size()) {
+                    break;
+                }
+                // increase the search granularity
+                partitionNum = Math.min(partitionNum * 2, knownToReproduceBugStatements.size());
+            }
+        }
 
         System.out.println("Reduced query:");
         printQueries(knownToReproduceBugStatements);
     }
 
     private List<Query<C>> tryReduction(G state, // NOPMD
             Reproducer<G> reproducer, G newGlobalState, List<Query<C>> knownToReproduceBugStatements,
-            BiFunction<List<Query<C>>, Integer, Boolean> reductionOperation) throws Exception {
+            TriFunction<List<Query<C>>, Integer, Integer, Boolean> reductionOperation) throws Exception {
 
         List<Query<C>> statements = knownToReproduceBugStatements;
-        for (int i = 0; i < statements.size(); i++) {
+
+        int start = 0;
+        int subLength = statements.size() / partitionNum;
+        while (start < statements.size()) {
+            // newStatements = candidate[:start] + candidate[start+subLength:]
+            // in other word, remove [start, start+subLength) from candidates
```

Comment:
I am confused by this comment. From lines L46-L53, we actually add [start, start + subLength] to candidate statements. Do I get your idea correctly?

Reply:
```
if (i >= s + l || i < s) {
      removed.add(candidateStatements.get(i));
}
candidateStatements.clear();
candidateStatements.addAll(removed);
```

It's my fault to name it too casually.  We keep the [: start) and [start+subLength:] actually.